### PR TITLE
Ability to link to external Coronavirus announcements with optional date

### PR DESCRIPTION
**Pull request by kevindew**
Trello: https://trello.com/c/CRwjuyqN/278-allow-external-links-from-announcements

This PR:

- Converts `Coronavirus::Announcement` to use a url field instead of path
- Adds an `AbsolutePathOrHttpsUrlValidator` and applies it `Coronavirus::Annoucement#url`
- Refactors the handling of the `published_at` so that it can be an optional field that can raise errors for partial dates
- Adds a helper that can strip `https://www.gov.uk` prefixes from URL's so absolute GOV.UK links are presented to the Publishing API as absolute paths
- Updates copy on the form (Nikin and Helen have reviewed it)

Further details are in the commits.

The form looks like so:

Screenshot 2021-03-29 at 10 59 18

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

File: app/models/coronavirus/announcement.rb [hunk 1]
```diff
@@ -18,13 +36,20 @@ def set_parent_positions
...
+  rescue ArgumentError
+    nil
   end
 
-  def valid_year?
-    published_at.past? && published_at.year > 1950
+  def valid_published_at
+    if published_at_hash && !published_at
+      errors.add(:published_at, "must be a valid date")
+    elsif published_at && published_at.year < 2000
```
Comment (leenagupte) [inline app/models/coronavirus/announcement.rb:49]: Nitpick: could this also be `published_at&.year < 2000` to be consistent with the clause below?

File: app/views/coronavirus/announcements/_form.html.erb [hunk 2]
```diff
@@ -35,17 +35,17 @@
     {
       name: "day",
       width: 2,
-      value: @announcement.published_at&.day
+      value: @announcement.published_at&.day || params.dig("announcement", "published_at", "day"),
```
Comment (leenagupte) [inline app/views/coronavirus/announcements/_form.html.erb:39]: Nice!

File: spec/models/coronavirus/announcement_spec.rb [hunk 3]
```diff
@@ -82,4 +91,29 @@
...
+      announcement.published_at = { "day" => "1", "month" => "13", "year" => "2020" }
+      expect(announcement.published_at).to be_nil
+    end
+
+    it "sets published_at to nil for an empty date" do
+      announcement.published_at = { "day" => "", "month" => "", "year" => "" }
+      expect(announcement.published_at).to be_nil
+    end
+
+    it "can still accept published_at as a time" do
```
Comment (leenagupte) [inline spec/models/coronavirus/announcement_spec.rb:113]: I'm not sure why we need this test?

File: app/helpers/govuk_url_helper.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+module GovukUrlHelper
+  def remove_govuk_from_url(url)
+    url.gsub(%r{\Ahttps://www\.gov\.uk}, "")
```
Comment (leenagupte) [inline app/helpers/govuk_url_helper.rb:3]: This is nice! 💯

**Review by leenagupte — COMMENTED** [review]
This looks really good! I've had a little bit of trouble testing it locally, but the code looks good. I especially love the new URL validator and GOV.UK url helper 💯 

Most of the comments are questions asking for clarification or minor non-blocking suggestions. The only thing holding back and approval is the catch-all test in the model. Once that's broken down this will be ready to go.

File: app/validators/absolute_path_or_https_url_validator.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class AbsolutePathOrHttpsUrlValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    uri = URI.parse(value)
+
+    absolute_https = uri.absolute? && uri.scheme == "https"
```
[the remark this answers is not in the document]
Comment (kevindew) [reply app/validators/absolute_path_or_https_url_validator.rb:5]: Sure, I can add something to the commit message for that. I figured there wasn't a scenario where we'd expect a content designer to link to a non HTTPS URL.

File: app/validators/absolute_path_or_https_url_validator.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    absolute_path = uri.relative? && uri.path.starts_with?("/") && value.starts_with?(uri.path)
```
[the remark this answers is not in the document]
Comment (kevindew) [reply app/validators/absolute_path_or_https_url_validator.rb:6]: It prevents complex relative URI's being included. For example without this check "//www.example.com/test" would pass validation.

File: app/models/coronavirus/announcement.rb [hunk 7]
```diff
@@ -2,9 +2,9 @@ class Coronavirus::Announcement < ApplicationRecord
   self.table_name = "coronavirus_announcements"
 
   belongs_to :page, foreign_key: "coronavirus_page_id", optional: false
-  validates :title, :url, presence: true
+  validates :title, presence: true
+  validates :url, presence: true, absolute_path_or_https_url: { allow_blank: true }
```
[the remark this answers is not in the document]
Comment (kevindew) [reply app/models/coronavirus/announcement.rb:6]: Yup, that's correct. it's a global option for validators.

File: app/models/coronavirus/announcement.rb (hunk 1 above)
Comment (kevindew) [reply app/models/coronavirus/announcement.rb:47]: Sadly not. With the safe navigation operator you end up risking `nil < 2000`  which risks a `NoMethodError (undefined method `<' for nil:NilClass)` exception.

File: spec/models/coronavirus/announcement_spec.rb (hunk 3 above)
Comment (kevindew) [reply spec/models/coronavirus/announcement_spec.rb:113]: I put this in as it's the only unit test that executes the https://github.com/alphagov/collections-publisher/pull/1335/files#diff-96d56b8391c8046c6b0d653114f673cfddced5572fe43faeecf16c959fe2b769R12-R16 path of code. I'm not sure if it's tested implicitly elsewhere in the codebase though.

File: spec/models/coronavirus/announcement_spec.rb [hunk 8]
```diff
@@ -1,34 +1,32 @@
...
-    it "requires title" do
-      announcement.title = ""
+    it { should validate_presence_of(:title) }
+    it { should validate_presence_of(:url) }
 
-      expect(announcement).not_to be_valid
-      expect(announcement.errors).to have_key(:title)
-    end
+    describe "url validations" do
+      it "validates url is an absolute path or https" do
```
[the remark this answers is not in the document]
Comment (kevindew) [reply spec/models/coronavirus/announcement_spec.rb:11]: Ah this was a real pain, it seemed a bit pointless to have a whole bunch of tests for this as they don't test any of announcement unit but instead test the validator (which is already tested). 

I'm going to have a little think about whether there is a different technique I can use.

[the remark this answers is not in the document]
Comment (kevindew) [reply spec/models/coronavirus/announcement_spec.rb]: I've gone with a shoulda matcher approach which seems less verbose and avoids having to retest everything.

**Comment by kevindew** [comment]
@leenagupte Thank you for the review. I've marked all the comments that didn't require further action as resolved and I'll await your next look at this.

**Review by leenagupte — APPROVED** [review]
Looks good! 👍 

There's one comment about a commit message, but it's not a blocker.

File: spec/models/coronavirus/announcement_spec.rb [hunk 9]
```diff
@@ -1,34 +1,21 @@
...
-      announcement.url = ""
+    it { should validate_presence_of(:title) }
+    it { should validate_presence_of(:url) }
 
-      expect(announcement).not_to be_valid
-      expect(announcement.errors).to have_key(:url)
-    end
+    describe "url validations" do
+      it { should allow_values("/path", "https://example.com").for(:url) }
+      it { should_not allow_values("not a url").for(:url) }
```
[the remark this answers is not in the document]
Comment (kevindew) [reply spec/models/coronavirus/announcement_spec.rb:12]: Thanks - I'll fix that before merging. I'll just give this a deploy to integration so Helen can give it a test run.

**Pull request merged**